cacher/cacher.go

```diff
@@ -47,50 +92,76 @@ func GetFromURLOrDefaultCache(req *http.Request, cacheDir string) (string, error
 	return result, err
 }
 
-// GetFromURLOrDefaultCacheWithCallback downloads file into cacheDir and passes the content to the onData callback
+// GetFromURLOrDefaultCacheWithCallback is a function maintained for compatibility.
+// The required httpClient is created on behalf of the caller.
+// This means some of metrics registered may be less meaningful.
+func GetFromURLOrDefaultCacheWithCallback(req *http.Request, cacheDir string, onData func([]byte) error) error {
+	httpClient := &promhttp.Client{
+		Client:     http.DefaultClient,
+		Registerer: prometheus.DefaultRegisterer,
+	}
+
+	return GetFromURLOrDefaultCacheWithCallbackAndClient(req, cacheDir, onData, httpClient)
+}
+
+// GetFromURLOrDefaultCacheWithCallbackAndClient downloads file into cacheDir and passes the content to the onData callback
 //
 // If the URL is not available, attempt to read the previous response from cache
 //
 // The onData function should process the data (e.g. parsing JSON) and return error on failure. It may be called a
 // second time to process cache if the data from remote URL cannot be processed.
 //
+// The httpClient is a wrapper for the standard net/http client which registers some basic metrics.
+// We do not register errors here because they would not be meaningful.
+//
 // The function only returns remote error if both downloading from the URL and reading from existing cache fail,
 // cache-related error is only logged, not reported.
-func GetFromURLOrDefaultCacheWithCallback(req *http.Request, cacheDir string, onData func([]byte) error) error {
+func GetFromURLOrDefaultCacheWithCallbackAndClient(req *http.Request, cacheDir string, onData func([]byte) error, httpClient *promhttp.Client) error {
```

Comment:
1. Why is this function part of the public API of this package? I imagine it won't be called directly from anywhere else so it could be private.
2. What is the purpose of the function overall? The only thing going on in the `GetFromURLOrDefaultCacheWithCallback` is that an http client is created, and then this function is called. Would it make sense to just merge this function into `GetFromURLOrDefaultCacheWithCallback` (as an alternative to making it a private function, see point above)?